### PR TITLE
Add support for Safeframe shrink to Doubleclick safeframe

### PR DESCRIPTION
**Pull request by bradfrizzell**
After this PR, AMP GPT safeframe will support a new shrink request, and will also now pass initial slot size as metadata. 

# IMPORTANT NOTE: This will not take effect until changes are live in ext.js to allow this message to be used within a safeframe. These changes are not yet available there.

### CODE REVIEW

File: extensions/amp-ad-network-doubleclick-impl/0.1/safeframe-host.js [hunk 1]
```diff
@@ -548,6 +555,22 @@ export class SafeframeHostApi {
...
+   */
+  handleShrinkRequest_(payload) {
+    if (!this.isRegistered_) {
+      return;
+    }
+    const shrinkHeight = Number(this.iframe_.height) -
+          payload['shrink_b'] + payload['shrink_t'];
+    const shrinkWidth = Number(this.iframe_.width) -
+          payload['shrink_r'] + payload['shrink_l'];
+    this.resizeAmpAdAndSafeframe(shrinkHeight, shrinkWidth,
```
Comment (keithwrightbos) [inline extensions/amp-ad-network-doubleclick-impl/0.1/safeframe-host.js]: What happens if shrink payload values are not present or numbers?  What about passing a size that's actually larger than the creative or is negative?

File: extensions/amp-ad-network-doubleclick-impl/0.1/test/test-doubleclick-safeframe.js [hunk 2]
```diff
@@ -649,5 +657,48 @@ describes.realWin('DoubleClick Fast Fetch - Safeframe', realWinConfig, env => {
         expect(resizeAmpAdAndSafeframeSpy).to.be.calledOnce;
       });
     });
+
+    /**
+     * Send a message requesting a shrink.
+     */
```
Comment (keithwrightbos) [inline extensions/amp-ad-network-doubleclick-impl/0.1/test/test-doubleclick-safeframe.js]: Add params to doc

File: extensions/amp-ad-network-doubleclick-impl/0.1/test/test-doubleclick-safeframe.js [hunk 3]
```diff
@@ -649,5 +657,48 @@ describes.realWin('DoubleClick Fast Fetch - Safeframe', realWinConfig, env => {
...
+    function sendShrinkMessage(height, width) {
+      const shrinkMessage = {};
+      shrinkMessage[MESSAGE_FIELDS.CHANNEL] = safeframeHost.channel;
+      shrinkMessage[MESSAGE_FIELDS.ENDPOINT_IDENTITY] = 1;
+      shrinkMessage[MESSAGE_FIELDS.SERVICE] = SERVICE.SHRINK_REQUEST;
+      shrinkMessage[MESSAGE_FIELDS.PAYLOAD] = JSON.stringify({
+        'uid': 0.623462509818004,
```
Comment (keithwrightbos) [inline extensions/amp-ad-network-doubleclick-impl/0.1/test/test-doubleclick-safeframe.js]: Where does this uid value come from?

File: extensions/amp-ad-network-doubleclick-impl/0.1/safeframe-host.js (hunk 1 above)
Comment (bradfrizzell) [reply extensions/amp-ad-network-doubleclick-impl/0.1/safeframe-host.js]: Done

File: extensions/amp-ad-network-doubleclick-impl/0.1/test/test-doubleclick-safeframe.js (hunk 2 above)
Comment (bradfrizzell) [reply extensions/amp-ad-network-doubleclick-impl/0.1/test/test-doubleclick-safeframe.js]: done

File: extensions/amp-ad-network-doubleclick-impl/0.1/test/test-doubleclick-safeframe.js (hunk 3 above)
Comment (bradfrizzell) [reply extensions/amp-ad-network-doubleclick-impl/0.1/test/test-doubleclick-safeframe.js]: Just sending a random number, which is what sf does. Although I guess this isn't really random as it's hardcoded, but it doesn't matter what it is .

File: extensions/amp-ad-network-doubleclick-impl/0.1/safeframe-host.js [hunk 4]
```diff
@@ -548,6 +557,28 @@ export class SafeframeHostApi {
...
+      return;
+    }
+    const shrinkHeight = Number(this.iframe_.height) -
+          payload['shrink_b'] + payload['shrink_t'];
+    const shrinkWidth = Number(this.iframe_.width) -
+          payload['shrink_r'] + payload['shrink_l'];
+    // Make sure we are actually shrinking here.
+    if (shrinkWidth > this.creativeSize_.width ||
+        shrinkHeight > this.creativeSize_.height) {
+      return;
```
Comment (keithwrightbos) [inline extensions/amp-ad-network-doubleclick-impl/0.1/safeframe-host.js]: log?  Does this handle the NaN case?

File: extensions/amp-ad-network-doubleclick-impl/0.1/safeframe-host.js [hunk 5]
```diff
@@ -578,15 +609,16 @@ export class SafeframeHostApi {
    * @param {number} height
    * @param {number} width
    * @param {string} messageType
-   * @param {boolean=} optIsCollapse
+   * @param {boolean=} optIsReducing True for collapse & shrink requests.
    */
-  resizeAmpAdAndSafeframe(height, width, messageType, optIsCollapse) {
+  resizeAmpAdAndSafeframe(height, width, messageType, optIsReducing) {
+    const isCollapsed = messageType == SERVICE.COLLAPSE_RESPONSE;
```
Comment (keithwrightbos) [inline extensions/amp-ad-network-doubleclick-impl/0.1/safeframe-host.js]: Do we need optIsReducing at all?  Can't we always just look at the messageType?

Comment (bradfrizzell) [reply extensions/amp-ad-network-doubleclick-impl/0.1/safeframe-host.js]: done

File: extensions/amp-ad-network-doubleclick-impl/0.1/safeframe-host.js [hunk 6]
```diff
@@ -468,7 +480,10 @@ export class SafeframeHostApi {
     if ((payload['push'] && !this.expandByPush_) ||
```
Comment (keithwrightbos) [inline extensions/amp-ad-network-doubleclick-impl/0.1/safeframe-host.js]: This is hard to parse.  Could we break it into separate checks or at least add some comments?

File: extensions/amp-ad-network-doubleclick-impl/0.1/safeframe-host.js [hunk 7]
```diff
@@ -544,10 +559,38 @@ export class SafeframeHostApi {
       this.resizeSafeframe(height, width, !!optIsCollapse, messageType);
     } else {
       this.resizeAmpAdAndSafeframe(
```
Comment (keithwrightbos) [inline extensions/amp-ad-network-doubleclick-impl/0.1/safeframe-host.js]: Can this be on 1 line?

File: extensions/amp-ad-network-doubleclick-impl/0.1/safeframe-host.js [hunk 8]
```diff
@@ -544,10 +559,38 @@ export class SafeframeHostApi {
...
+        isNaN(payload['shrink_r']) || isNaN(payload['shrink_l'])) {
+      dev().error(TAG, 'Invalid shrink values.');
+      return;
+    }
+    const shrinkHeight = Number(this.iframe_.height) -
+          payload['shrink_b'] + payload['shrink_t'];
+    const shrinkWidth = Number(this.iframe_.width) -
+          payload['shrink_r'] + payload['shrink_l'];
+    // Make sure we are actually shrinking here.
+    if (shrinkWidth > this.creativeSize_.width ||
```
Comment (keithwrightbos) [inline extensions/amp-ad-network-doubleclick-impl/0.1/safeframe-host.js]: You can do the nan checks here as number + NaN = NaN. E.g:

if (isNaN(shrinkWidth) || isNaN(shrinkHeight) || ..) return;

File: extensions/amp-ad-network-doubleclick-impl/0.1/safeframe-host.js [hunk 9]
```diff
@@ -578,15 +621,17 @@ export class SafeframeHostApi {
...
+    const isCollapsed = messageType == SERVICE.COLLAPSE_RESPONSE;
+    const isReducing = messageType == SERVICE.COLLAPSE_RESPONSE ||
+          messageType == SERVICE.SHRINK_RESPONSE;
     // First, attempt to resize the Amp-Ad that is the parent of the
     // safeframe
     this.baseInstance_.attemptChangeSize(height, width).then(() => {
       // If this resize succeeded, we always resize the safeframe.
       // resizeSafeframe also sends the resize response.
-      this.resizeSafeframe(height, width, !!optIsCollapse, messageType);
+      this.resizeSafeframe(height, width, isCollapsed, messageType);
```
Comment (keithwrightbos) [inline extensions/amp-ad-network-doubleclick-impl/0.1/safeframe-host.js]: Why do we need isCollapsed if we're passing messageType?

File: extensions/amp-ad-network-doubleclick-impl/0.1/safeframe-host.js [hunk 10]
```diff
@@ -598,13 +643,14 @@ export class SafeframeHostApi {
...
       this.baseInstance_.getResource().resetPendingChangeSize();
-      if (optIsCollapse) {
-        // If this is a collapse request, then even if resizing
+      if (isReducing) {
+        // If this is a collapse or shrink request, then even if resizing
         // the amp-ad failed, still resize the iframe.
         // resizeSafeframe also sends the resize response.
-        this.resizeSafeframe(height, width, !!optIsCollapse, messageType);
+        // Only register as collapsed if explicitly a collapse request.
+        this.resizeSafeframe(height, width, isCollapsed, messageType);
```
Comment (keithwrightbos) [inline extensions/amp-ad-network-doubleclick-impl/0.1/safeframe-host.js]: here too

File: extensions/amp-ad-network-doubleclick-impl/0.1/safeframe-host.js [hunk 11]
```diff
@@ -598,13 +643,14 @@ export class SafeframeHostApi {
       // to execute upon the next user interaction. We don't want
       // that for safeframe, so we reset it here.
       this.baseInstance_.getResource().resetPendingChangeSize();
-      if (optIsCollapse) {
-        // If this is a collapse request, then even if resizing
+      if (isReducing) {
```
Comment (keithwrightbos) [inline extensions/amp-ad-network-doubleclick-impl/0.1/safeframe-host.js]: No need to declare isReducing in advance, can just put logic here

File: extensions/amp-ad-network-doubleclick-impl/0.1/safeframe-host.js (hunk 10 above)
Comment (bradfrizzell) [reply extensions/amp-ad-network-doubleclick-impl/0.1/safeframe-host.js]: done

File: extensions/amp-ad-network-doubleclick-impl/0.1/safeframe-host.js (hunk 11 above)
Comment (bradfrizzell) [reply extensions/amp-ad-network-doubleclick-impl/0.1/safeframe-host.js]: done

File: extensions/amp-ad-network-doubleclick-impl/0.1/safeframe-host.js (hunk 9 above)
Comment (bradfrizzell) [reply extensions/amp-ad-network-doubleclick-impl/0.1/safeframe-host.js]: removed.

File: extensions/amp-ad-network-doubleclick-impl/0.1/safeframe-host.js (hunk 8 above)
Comment (bradfrizzell) [reply extensions/amp-ad-network-doubleclick-impl/0.1/safeframe-host.js]: done

File: extensions/amp-ad-network-doubleclick-impl/0.1/safeframe-host.js (hunk 7 above)
Comment (bradfrizzell) [reply extensions/amp-ad-network-doubleclick-impl/0.1/safeframe-host.js]: done
